File: Leetcode/589_N-ary_Tree_Preorder_Traversal.js
/**
 * // Definition for a Node.
 * function Node(val, children) {
 *    this.val = val;
 *    this.children = children;
 * };
 */
/**
 * @param {Node} root
 * @return {number[]}
 */
var preorder = function (root, result = []) {
    if (root) {
        result.push(root.val);

        root.children.forEach(node => {
            preorder(node, result);
        })

        return result;
    }
    return [];
};

File: Leetcode/0589/589_N-ary_Tree_Preorder_Traversal.cpp
/*
// Definition for a Node.
class Node {
public:
    int val;
    vector<Node*> children;

    Node() {}

    Node(int _val, vector<Node*> _children) {
        val = _val;
        children = _children;
    }
};
*/
class Solution {
public:
    vector<int> ans;
    vector<int> preorder(Node* root) {
        if(root != nullptr)
        {
            ans.push_back(root->val);
            
            for(const auto node : root->children)
            {
                preorder(node);    
            }
        }
        return ans;
    }
};